pkg/storage/mvcc.go

```diff
@@ -3705,6 +3706,216 @@ func MVCCDeleteRange(
+// MVCCPredicateDeleteRangePointTombstones returns a WriteTooOldError without a
+// resumeSpan, even if tombstones were already written to disk. To resolve, the
+// caller should retry the call at a higher timestamp, assuming they have the
+// appropriate level of isolation (e.g. the span covers an offline table, in
+// the case of IMPORT rollbacks).
+//
+// An example of how this works: Issuing DeleteRange[a,e)@3 with
+// Predicate{StartTime=1} on the following keys would issue tombstones at a@3,
+// b@3, and d@3.
+//
+// t3
+// t2 a2 b2    d2 e2
+// t1    b1 c1
+//
+//	a  b  c  d  e
+func MVCCPredicateDeleteRangePointTombstones(
+	ctx context.Context,
+	rw ReadWriter,
+	ms *enginepb.MVCCStats,
+	startKey, endKey roachpb.Key,
+	endTime hlc.Timestamp,
+	localTimestamp hlc.ClockTimestamp,
+	predicates kvpb.DeleteRangePredicates,
+	maxBatchSize int64,
+	maxBatchByteSize int64,
+	maxLockConflicts int64,
+	targetLockConflictBytes int64,
+) (*roachpb.Span, error) {
+	if endTime.IsEmpty() {
+		return nil, errors.AssertionFailedf("MVCCPredicateDeleteRange expects non-empty endTime")
+	}
+	if ms == nil {
+		return nil, errors.AssertionFailedf(
+			"MVCCStats passed in to MVCCPredicateDeleteRange must be non-nil to ensure proper stats" +
+				" computation during Delete operations")
+	}
+
+	// Check for any overlapping locks, and return them to be resolved.
+	if locks, err := ScanLocks(ctx, rw, startKey, endKey, maxLockConflicts, targetLockConflictBytes); err != nil {
+		return nil, err
+	} else if len(locks) > 0 {
+		return nil, &kvpb.LockConflictError{Locks: locks}
+	}
+
+	// Create some reusable machinery for flushing a run with point tombstones
+	// that is typically used in a single MVCCPut call.
+	pointTombstoneIter, err := newMVCCIterator(
+		ctx, rw, endTime, false /* rangeKeyMasking */, true, /* noInterleavedIntents */
+		IterOptions{
+			KeyTypes:     IterKeyTypePointsAndRanges,
+			Prefix:       true,
+			ReadCategory: fs.BatchEvalReadCategory,
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer pointTombstoneIter.Close()
+
+	ltScanner, err := newLockTableKeyScanner(
+		ctx, rw, uuid.UUID{} /* txnID */, lock.Intent,
+		maxLockConflicts, targetLockConflictBytes, fs.BatchEvalReadCategory,
+	)
+	if err != nil {
+		return nil, err
+	}
+	defer ltScanner.close()
+
+	pointTombstoneBuf := newPutBuffer()
+	defer pointTombstoneBuf.release()
+
+	deleteKey := func(key roachpb.Key) error {
+		// Use Point tombstones
+		_, acq, err := mvccPutInternal(
+			ctx, rw, pointTombstoneIter, ltScanner, key, endTime, noValue, pointTombstoneBuf,
+			nil, MVCCWriteOptions{
+				LocalTimestamp: localTimestamp, Stats: ms, Category: fs.BatchEvalReadCategory},
+		)
+		if err != nil {
+			return err
+		}
+		if !acq.Empty() {
+			log.Dev.Fatal(ctx, "expected empty lock acquisition for non-transactional point delete")
+		}
+		return nil
+	}
+
+	iter, err := newMVCCIterator(ctx, rw, endTime, true, true, IterOptions{
```

Comment:
MVCCPredicateDeleteRanges uses an MVCCIncrementalIterator. Might be worth a comment on why we aren't using that here.